### ISSUE TITLE
Running tuple definitions in parallel

### ISSUE THREAD

**Issue by gchauras** [halide/Halide#624]
Is there any way I can specify to compute the different Tuple elements in parallel.
f(x) = Tuple(g(x), h(x))
f.parallel(x)
But the two Tuple elements are serialized with the above.

**Comment by abadams** [comment]
We're not good at task-parallelism, but if the Tuple elements are the same
type, you could try something like:

f(x, i) = select(i == 0, g(x), h(x));
g.compute_at(f, i);
h.compute_at(f, i);
f.bound(i, 0, 2).parallel(i);

There's a redundant memcpy there though. If they're different types you
could use the evil construction:

f(x) = Tuple(undef<int>(), undef<float>());
RDom r(0, 2);
f(x) = Tuple(select(r == 0, g(x), undef<int>()),
                  select(r == 1, undef<float>(), h(x)));
g.compute_at(f, r);
h.compute_at(f, r);
f.update().reorder(x, r).allow_race_conditions().parallel(r);

I think what you'd really like to write is something like:
g.compute_root();
h.compute_in_parallel_with(g);

Where compute_in_parallel_with puts h in a task-parallel block alongside
the computation of g.


On Sun, Jan 11, 2015 at 12:43 PM, Gaurav Chaurasia <notifications@github.com

**Comment by gchauras** [comment]
I know about putting an outermost index and using select. I was wondering if there was any support for task parallelism. The second one I had not thought of. I will keep in my arsenal. Since I am working with GPU parallelism, the first option is really the only choice. Even if you supported task parallelism, providing this for GPU would be nasty -- this will have to be mapped to thread_id_x. Thanks anyway.

**Comment by ashishUthama** [comment]
I think a compute_in_parallel_with would be nice to have. 
(I am struggling to adopt and schedule the 'evil construction' trick).
